velox/functions/prestosql/aggregates/PrestoHasher.cpp

```diff
@@ -0,0 +1,208 @@
+}
+
+template <typename T>
+FOLLY_ALWAYS_INLINE int64_t
+getHashForIntTypes(const DecodedVector& vector, vector_size_t row) {
+  auto value = vector.valueAt<T>(row);
+  return getHashForIntTypes(value);
+}
+
+FOLLY_ALWAYS_INLINE int64_t
+getHashForVarCharTypes(const DecodedVector& vector, vector_size_t row) {
+  auto input = vector.valueAt<StringView>(row);
+  int64_t hash = XXH64(input.data(), input.size(), 0);
+  return hash;
+}
+} // namespace
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::BIGINT>(vector_size_t row) {
+  return getHashForIntTypes<int64_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::INTEGER>(vector_size_t row) {
+  return getHashForIntTypes<int32_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::SMALLINT>(vector_size_t row) {
+  return getHashForIntTypes<int16_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::TINYINT>(vector_size_t row) {
+  return getHashForIntTypes<int8_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::DOUBLE>(vector_size_t row) {
+  auto value = vector_->valueAt<double>(row);
+  int64_t u = *reinterpret_cast<int64_t*>(&value);
+  return getHashForIntTypes(u);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::DATE>(vector_size_t row) {
+  return getHashForIntTypes<int32_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::VARCHAR>(vector_size_t row) {
+  return getHashForVarCharTypes(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::VARBINARY>(vector_size_t row) {
+  return getHashForVarCharTypes(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::REAL>(vector_size_t row) {
+  auto value = vector_->valueAt<double>(row);
+  int32_t u = *reinterpret_cast<int32_t*>(&value);
+  return getHashForIntTypes(u);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::TIMESTAMP>(vector_size_t row) {
+  return getHashForIntTypes<int64_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::BOOLEAN>(vector_size_t row) {
+  auto value = vector_->valueAt<bool>(row);
+  return value ? 1231 : 1237;
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::ARRAY>(vector_size_t row) {
+  int64_t hash = 0;
+  auto baseArray = vector_->base()->as<velox::ArrayVector>();
+  auto rawSizes = baseArray->rawSizes();
+  auto rawOffsets = baseArray->rawOffsets();
+  auto indices = vector_->indices();
+
+  auto size = rawSizes[indices[row]];
+  auto offset = rawOffsets[indices[row]];
+
+  if (children_.empty()) {
+    children_.reserve(1);
+    SelectivityVector elementRows(baseArray->elements()->size());
+    auto elements =
+        std::make_shared<DecodedVector>(*baseArray->elements(), elementRows);
+    children_.push_back(std::make_unique<PrestoHasher>(elements));
```

Comment:
This code repeats multiple times. Consider extracting a helper method to adding a constructor to PrestoHasher that takes a vector.

```
children_.push_back(makePrestoHasher(baseArray->elements()));
```

or

```
children_.push_back(std::make_unique<PrestoHasher>(baseArray->elements());
```